types/wlr_relative_pointer_v1.c

```diff
@@ -0,0 +1,234 @@
+	wlr_relative_pointer_manager_v1_destroy(relative_pointer_manager);
+}
+
+
+/**
+ * Implementations
+ */
+
+static const struct zwp_relative_pointer_manager_v1_interface relative_pointer_manager_v1_impl = {
+	.destroy = relative_pointer_manager_v1_handle_destroy,
+	.get_relative_pointer = relative_pointer_manager_v1_handle_get_relative_pointer,
+};
+
+
+static const struct zwp_relative_pointer_v1_interface relative_pointer_v1_impl = {
+	.destroy = relative_pointer_v1_handle_destroy,
+};
+
+
+/**
+ * Public functions
+ */
+
+struct wlr_relative_pointer_manager_v1 *wlr_relative_pointer_manager_v1_create(struct wl_display *display) {
+	struct wlr_relative_pointer_manager_v1 *relative_pointer_manager =
+		calloc(1, sizeof(struct wlr_relative_pointer_manager_v1));
+
+	if (relative_pointer_manager == NULL) {
+		return NULL;
+	}
+
+	wl_list_init(&relative_pointer_manager->resources);
+	wl_list_init(&relative_pointer_manager->relative_pointers);
+
+	relative_pointer_manager->global = wl_global_create(display,
+		&zwp_relative_pointer_manager_v1_interface, RELATIVE_POINTER_MANAGER_VERSION,
+		relative_pointer_manager, relative_pointer_manager_v1_bind);
+
+	if (relative_pointer_manager->global == NULL) {
+		free(relative_pointer_manager);
+		return NULL;
+	}
+
+	wl_signal_init(&relative_pointer_manager->events.destroy);
+	wl_signal_init(&relative_pointer_manager->events.new_relative_pointer);
+
+	relative_pointer_manager->display_destroy_listener.notify = handle_display_destroy;
+	wl_display_add_destroy_listener(display, &relative_pointer_manager->display_destroy_listener);
+
+	wlr_log(WLR_DEBUG, "relative_pointer_v1 manager created");
+
+	return relative_pointer_manager;
+}
+
+
+void wlr_relative_pointer_manager_v1_destroy(struct wlr_relative_pointer_manager_v1 *relative_pointer_manager) {
+	if (relative_pointer_manager == NULL) {
+		return;
+	}
+
+	wlr_signal_emit_safe(&relative_pointer_manager->events.destroy, relative_pointer_manager);
+	wl_list_remove(&relative_pointer_manager->display_destroy_listener.link);
+
+	struct wl_resource *resource;
+	struct wl_resource *tmp_resource;
+	wl_resource_for_each_safe(resource, tmp_resource, &relative_pointer_manager->resources) {
+		wl_resource_destroy(resource);
+	}
+
+	wl_global_destroy(relative_pointer_manager->global);
+	free(relative_pointer_manager);
+
+	wlr_log(WLR_DEBUG, "relative_pointer_v1 manager destroyed");
+}
+
+
+void wlr_relative_pointer_v1_send_relative_motion(struct wlr_relative_pointer_v1 *relative_pointer,
+		uint64_t time, double dx, double dy,
+		double dx_unaccel, double dy_unaccel) {
+	zwp_relative_pointer_v1_send_relative_motion(relative_pointer->resource,
+		(uint32_t)(time >> 32), (uint32_t)time,
+		wl_fixed_from_double(dx), wl_fixed_from_double(dy),
+		wl_fixed_from_double(dx_unaccel), wl_fixed_from_double(dy_unaccel));
+}
+
+
+struct wlr_relative_pointer_v1 *wlr_relative_pointer_v1_from_resource(struct wl_resource *resource) {
+	return relative_pointer_from_resource(resource);
```

Comment:
We can just rename `relative_pointer_from_resource` to `wlr_relative_pointer_v1_from_resource`